[10.0][ADD] stock_available_product_expiry

rousseldenis opened the pull request.

rousseldenis committed:
[10.0][ADD] stock_availbale_product_expiry

rousseldenis commented:
As it breaks stock_available modules, I will add a config parameter to enable this feature or not.

rousseldenis commented:
Fix commits squashed. Please don't squash these.

lmignon reviewed (approved):
@rousseldenis Thank you for this useful addon (IMO)
:+1: (Code review /functional)

pedrobaeza reviewed (commented):
Shouldn't this be `stock_available_product_expiry` to follow the same terminology?

lmignon commented:
@pedrobaeza module renamed and commit squashed/fixed

pedrobaeza commented:
Setup files haven't been renamed

pedrobaeza commented:
Can you add also some tests for not dropping coverage?

lmignon commented:
I'm improving this PR. -> WIP

lmignon committed:
[IMP] stock_availabke_product_expiry: Add qty_expired field

lmignon committed:
[IMP] stock_available_product_expiry: Add new models to ease the scrapping of expired products

lmignon commented:
Works done. This PR is ready for reviews

lmignon committed:
[IMP] stock_available_product_expiry: Add missing index and improve tests coverage

lmignon committed:
[FIX] stock_available_product_expiry: outgoing_expired_qty is correctly computed
In the same time, the perfomances are improved and the code is now 2 times faster than in the previous version

lmignon commented:
PLZ don't squash commits on this PR

rousseldenis committed:
[IMP] stock_available_product_expiry: use _compute_quantities on product.template
In the same time fix in the domain computation on product.product:

rousseldenis committed:
[IMP] stock_available_product_expiry: Remove unused function in product.product

rousseldenis committed:
[IMP] stock_available_product_expiry: Add function to bypass expired quantities domain computation

rousseldenis committed:
[IMP] stock_available_product_expiry: Improve coverage

rousseldenis committed:
[IMP] stock_available_product_expiry: Avoid already reserved quants to be scrapped

If between the moment of scrap has been created and the moment of the scrap is confirmed, the concerned quants
are reserved, the picking could create negative quants. Correct that behaviour

rousseldenis committed:
[IMP] stock_available_product_expiry: Improve computation performance

lmignon commented on stock_available_product_expiry/models/stock_quant.py:
```diff
@@ -0,0 +1,18 @@
+
+    lot_id = fields.Many2one(
+        'stock.production.lot',
+        auto_join=True
```
@rousseldenis IMO the auto_join works only with required fields. Can you check that quants without lot and reservation_id are still visible with `auto_join=True`?

rousseldenis replied:
Ok, I'll check

rousseldenis replied:
Indeed, not visible... :-(

I try to find another way to compute

lmignon commented on stock_available_product_expiry/models/product_product.py:
```diff
@@ -107,16 +106,35 @@ def _compute_quantities(self):
             product.qty_expired -= product.outgoing_expired_qty
         return res
 
+    def _get_quants(self, quant_domain, quant_out_domain):
```
@rousseldenis You must be sure that the original value for `auto_join` on the field is restored even in case of exception....

```python
@api.model
@contextmanager
def _force_auto_join(self, model_field_list):
    """ Force  auto_join for each item (model_name, field_name) in model_field_list
    The original value are restored on exit
    """
    vals = {}
    for model_name, field_name in model_field_list:
        field = self.env[model_name]._fields[field_name]
        vals[field] = field.auto_join
        field.auto_join = True
    try:
        yield
    finally:
        for field, value in vals.iteritems():
            field.auto_join = value

def  _get_quants(self, quand_domain, quant_out_domain):
    stock_quant_obj = self.env['stock.quant']
    with self._force_auto_join([('stock.quant', 'lot_id')('stock.quant', 'reservation_id')]:
        quants = stock_quant_obj.with_context(lang='').read_group(
            quant_domain, ['product_id', 'qty'], ['product_id'])
        quants_out = stock_quant_obj.with_context(lang='').read_group(
            quant_out_domain, ['product_id', 'qty'], ['product_id'])
        return quants, quants_out
```

rousseldenis replied:
Done

rousseldenis committed:
[FIX] README, flake8, context

lmignon commented:
@rousseldenis Can you squash the two commits 'Imrove computation performance' PLZ

rousseldenis commented:
@lmignon Seems ready to be merged (or final review) ?

@pedrobaeza cc@eficent

rousseldenis commented:
@lreficent I think this one can be merged but missing review

LoisRForgeFlow commented:
I just rebuilt the runbot, it seems ok, but I'm not able to access :confused:

rousseldenis committed:
[IMP] stock_available_expiry: Add possibility to disable expired qties computation

gurneyalex commented on stock_available_product_expiry/models/product_product.py:
```diff
@@ -0,0 +1,179 @@
+    def action_open_expired_quants(self):
+        action = self.env.ref('stock.product_open_quants').read()[0]
+        domain = [
+            ('product_id', '=', self.id),
```
maybe change to ('product_id', 'in', self.ids) and this can get called from a list view

rousseldenis commented:
@gurneyalex Done

rousseldenis committed:
[10.0][FIX] stock_available_product_expiry: Fix domains, activate parameter

Take into account if removal date is set or not in domains
Activate the ir config parameter by default

rousseldenis committed:
[10.0][IMP] stock_available_product_expiry: New OCA conventions

rousseldenis commented:
@Cedric-Pigeon Can you approve this as you tested latest changes ?
